### DISCUSSION TITLE
Issues with Vehicle Envelope Configuration and Resource Allocation in openTCS v5.17

### DISCUSSION THREAD

**Discussion by csliao0524** [openTCS/opentcs#154] [Q&A] [answered]
Hi,

Now, I'm using openTCS v5.17 and have referenced the user's guide for setting up the vehicle envelope. Here's what I've done:

1. Configured the kernel for the route allocation mechanism.
![image](https://github.com/user-attachments/assets/22842681-7e28-4674-abe5-8dadde5fde50)

2. Set the envelope key 
(in my case, there are 2 vehicles with same type, e.g., M115, and therefore they have the same size, e.g., 800mm * 800mm).
![image](https://github.com/user-attachments/assets/9b81c576-8b23-40c0-a833-f2f8389b36e0)
![image](https://github.com/user-attachments/assets/2fb1b260-815d-4af7-ac19-8a96fae2503c)

3. Set the envelope values at all points and paths. 
I've given the same envelope setting for each point and path as follows:
![image](https://github.com/user-attachments/assets/7d5d0ef0-c445-4e04-91ca-ab8ab1cacdf8)

4. Results and problems
I expect that the simulated vehicle would occupy resource according to the envelope range. However, the actual situation is as follows:

- Initial position and resource allocation 
![image](https://github.com/user-attachments/assets/5944ad8e-7251-45c6-88ec-276d76e24b79)

-Moving vehicle-0001 from point-0001 to point-0002 
(I did not see it occupy multiple resource based on the envelope range during this process)
![image](https://github.com/user-attachments/assets/20b62712-909e-4e5c-aaea-7dbe38e440ae)

- Dispatched vehicle-0002 to Location-0002
(Encountered "TOrder is BEING_PROCESSED" but no action is taken (it doesn't seem to be a resource issue either?)
![image](https://github.com/user-attachments/assets/3357ba6c-c4b5-41a8-9286-1beb48510f6a)

Do you have any solutions or suggestions? Did I misunderstand or misconfigure something? Thank you for your response!

**Comment by csliao0524** [comment]
Hi maintainer:

To simplify the scenario (still using v5.17), as shown in the diagram: 
There is only 1 vehicle (envelope key = M115),  4 points with coordinates (0, 0), (400, 400), (500, 0), and (1500, 0), respectively. Each point and path element has vehicle envelopes set as [(0,0), (800,0), (800,800), (0,800), (0,0)] and I've also the set the kernel config kernelapp.vehicleResourceManagementType = LENGTH_IGNORED.

After several dispatches and observing the resource allocation, it seems that the vehicle envelope still has no effect. 
What could be the possible reasons for this? Additionally, how should the coordinates of the vehicle envelope be defined? Is it defined like how I drew it in the diagram?

IDLE at the Point-0001 
![image](https://github.com/user-attachments/assets/5bdc4265-6688-4c0c-b4ad-5a20922c008a)

Moving from Point-0001 to Point-0002
![image](https://github.com/user-attachments/assets/5aca27b7-5b02-485e-9129-2420f075841d)

**Reply by martingr** [reply]
You wrote:

>  Each point and path element has vehicle envelopes set as [(0,0), (800,0), (800,800), (0,800), (0,0)] 

With vehicle enveloples, the coordinates are absolute coordinates. Setting the vehilce envelope on all point and path elements to [(0,0), (800,0), (800,800), (0,800), (0,0)] means that they all describe/define _the same_ envelope / area in the plant model. Although this is not prohibited, you probably want different vehicle envelopes at every point/path. After all, vehicle envelopes at points/paths represent the area that a vehicle "occupies" when it allocates the respective points/paths.

What do you mean with:

> After several dispatches and observing the resource allocation, it seems that the vehicle envelope still has no effect.

What is your expectation regarding resource allocation?

**Reply by csliao0524** [reply]
For the example below
![image](https://github.com/user-attachments/assets/302deaef-041a-4ccf-be2a-f99f1e882204)

Under the setting of vehicle envelope [(0,0), (800,0), (800,800), (0,800), (0,0)]
I expect that when the vehicle at these places: Point-0001, Point-0001 --- Point-0002, Point-0002, the vehicle would also be allocated the resource of Point-0004, Point-0001 --- Point-0004 at the same time(even the coordinates of envelope are absolute coordinates)?

**Reply by martingr** [reply, marked as answer]
This is not how vehicle envelopes work. A vehicle envelope defined at a point primarily describes the area that a vehicle occupies when it allocates that point. The same applies to paths. See the section "Allocating resources" in the [user's guide] for details on how vehicle envelopes affect resource allocation.

In your example, the vehicle only allocates Point-0001 and thus also the area that is associated with that point's vehicle envelope. The existence of other points/paths in that area _will not_ result in the vehicle to automatically allocate the resources for these points/paths as well. However, openTCS _will_ prevent other vehicles to allocate resources where vehicle envelopes are defined that insertect with the vehicle envelope at Point-0001.

[user's guide]: https://www.opentcs.org/docs/6.0/user/opentcs-users-guide.html#_default_scheduler
